ChildProject/utils.py

```diff
@@ -45,6 +46,46 @@ def intersect_ranges(xs, ys):
         except StopIteration:
             return
 
+class TimeInterval:
+    def __init__(self, start : datetime, stop : datetime):
+        #remove the day/month/year component
+        self.start = start.replace(year=1900, month=1, day=1)
+        self.stop = stop.replace(year=1900, month=1, day=1)
+
+    def length(self):
+        return self.stop - self.start
+
+    def __repr__(self):
+        return "TimeInterval([{}, {}])".format(self.start, self.stop)
+    
+def time_intervals_intersect(ti1 : TimeInterval, ti2 : TimeInterval):
+    """
+    given 2 time intervals (those do not take in consideration days, only time in the day), return an array of new interval(s) representing the intersections of the original ones.
+    eg :
+        - time_intervals_intersect(TimeInterval(datetime(1900,1,1,8,57),datetime(1900,1,1,21,4)),TimeInterval(datetime(1900,1,1,10,36),datetime(1900,1,1,22,1))) => [TimeInterval(10:36 , 21:04)]
+        - time_intervals_intersect(TimeInterval(datetime(1900,1,1,8,57),datetime(1900,1,1,22,1)),TimeInterval(datetime(1900,1,1,21,4),datetime(1900,1,1,10,36))) => [TimeInterval(08:57 , 10:36),TimeInterval(21:04 , 22:01)]
+    """
+    #The calculation and boolean evaluation is done that way to optimize the process, those expressions were obtained using a Karnaugh table. Given the relations between the different start and ending times, the boolean relations used below gives the correct intervals
+    a = ti1.start <= ti1.stop
+    b = ti2.start <= ti2.stop
+    c = ti1.stop <= ti2.stop
+    d = ti1.start <= ti2.start
+    e = ti1.start <= ti2.stop
+    f = ti2.start <= ti1.stop
+    r = []
+    if c and (d and (not e or f) or not e and f) or d and not e and f : r = [TimeInterval(ti2.start,ti1.stop)]
+    elif not c and (d and (b or not a) or not a and b) or not a and b and d : r = [ti2]
+    elif not c and (not d and (not e and not f or e) or e and not f) or not d and e and not f : r = [TimeInterval(ti1.start,ti2.stop)]
+    elif c and (not d and (not a and not b or a) or a and not b) or a and not b and not d : r = [ti1]
+    # !! here the expression was simplified because previous statements should already have caught their relevant cases (ie this statement should always be last or changed)
+    elif not a and (not b or e) or d and e and f : r = [TimeInterval(ti1.start,ti2.stop),TimeInterval(ti2.start,ti1.stop)]
```

Comment:
Perhaps it should be commented on a little more ;)

Reply:
yes, probably... Added some line to at least have some info on what it does but I don't think explaining the whole process that lead to those boolean expression is really useful. Or we could change the whole function to work in a more comprehensible (and probably less optimized) way ?

Reply:
I found a solution that is a little bit more verbose but clearer in my opinion (I haven't tested it thoroughly though). 
Here is the dump of my Jupyter notebook:

```python
from datetime import datetime
```


```python
class TimeInterval:
    def __init__(self, start : datetime, stop : datetime):
        #remove the day/month/year component
        self.start = start.replace(year=1900, month=1, day=1)
        self.stop = stop.replace(year=1900, month=1, day=1)

    def length(self):
        return self.stop - self.start

    def __repr__(self):
        return "TimeInterval([{}, {}])".format(self.start, self.stop)
```


```python
# Utility functions
def overlap(onset, offset, target_onset, target_offset):
    return max(0, min(offset, target_offset) - max(onset, target_onset))

def overlaps(onset, offset, target_onset, target_offset):
    return bool(overlap(onset, offset, target_onset, target_offset))

def time_as_decimal(datetime_):
    return datetime_.hour + datetime_.minute / 60 + datetime_.second / 3600

def break_time(ti : TimeInterval):
    ti_ = []
    if ti.start < ti.stop:
        ti_.append((ti.start, ti.stop))
    else:
        # Break at midnight
        ti_.extend([(ti.start, datetime(1900,1,1,23,59)), 
                    (datetime(1900,1,1,0,0), ti.stop)])
    return ti_
```


```python
# Should break
break_time(TimeInterval(datetime(1900,1,1,21,4),datetime(1900,1,1,10,36)))
```




    [(datetime.datetime(1900, 1, 1, 21, 4), datetime.datetime(1900, 1, 1, 23, 59)),
     (datetime.datetime(1900, 1, 1, 0, 0), datetime.datetime(1900, 1, 1, 10, 36))]




```python
# Shouldn't break
break_time(TimeInterval(datetime(1900,1,1,8,57),datetime(1900,1,1,22,1)))
```




    [(datetime.datetime(1900, 1, 1, 8, 57), datetime.datetime(1900, 1, 1, 22, 1))]




```python
# Convert to decimal so as to be usable by overlap (maybe change this later)
time_as_decimal(datetime(1900,1,1,22,59,59))
```




    22.999722222222225




```python
def time_intervals_intersect(ti1 : TimeInterval, ti2 : TimeInterval):
    ti1_broken = break_time(ti1)
    ti2_broken = break_time(ti2)
    
    olaps = []
    for start1, stop1 in ti1_broken:
        st1 = time_as_decimal(start1)
        sp1 = time_as_decimal(stop1)
        
        for start2, stop2 in ti2_broken:
            st2=time_as_decimal(start2)
            sp2=time_as_decimal(stop2)

            if overlap(st1,sp1,st2,sp2):
                olaps.append((max(start1,start2), min(stop1,stop2)))
    return olaps
```


```python
time_intervals_intersect(TimeInterval(datetime(1900,1,1,8,57),datetime(1900,1,1,21,4)),TimeInterval(datetime(1900,1,1,10,36),datetime(1900,1,1,22,1)))
```




    [(datetime.datetime(1900, 1, 1, 10, 36), datetime.datetime(1900, 1, 1, 21, 4))]




```python
time_intervals_intersect(TimeInterval(datetime(1900,1,1,8,57),datetime(1900,1,1,22,1)),TimeInterval(datetime(1900,1,1,21,4),datetime(1900,1,1,10,36)))
```




    [(datetime.datetime(1900, 1, 1, 21, 4), datetime.datetime(1900, 1, 1, 22, 1)),
     (datetime.datetime(1900, 1, 1, 8, 57), datetime.datetime(1900, 1, 1, 10, 36))]

Reply:
What you could do too is use `time` instead of `datetime` with a date set to 1900  (which is a bit hacky in my opinion)

```
        self.start = time(start.hour, start.minute, start.second)
        self.stop = time(stop.hour, stop.minute, stop.second)
```